Do not focus custom fields in the create child dialog

cbliard opened the pull request:
# Ticket

PR #17625
Late review revealed an issue. This PR is a follow-up about it.

# What are you trying to accomplish?

In the create child dialog, displayable from the relations tab of a work package, custom fields using the autocompleter input field are getting the focus by default. This is visible only if the work package type has some specific kind of custom fields like `hierarchy`, `version` or `list`.

This is not correct: the field to focus is the subject field.

Also add a test to ensure that all different kind of custom fields are displayed in the create child dialog and no error is reported.

# What approach did you choose and why?

Pass in `focusDirectly: false` so that the focus is not stolen. This may benefit some other areas of the app.

# Merge checklist

- [X] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [X] Tested major browsers (Chrome, Firefox, Edge, ...)

cbliard committed:
Do not focus custom fields in the create child dialog

The field to focus is the subject field.

Also test that all custom fields are displayed in the create
child dialog.

Also updated the `:custom_field` factory for "hierarchy" to create the
root item using `CustomFields::Hierarchy::HierarchicalItemService.new`
by default. A factory should not created half-baked objects.

NobodysNightmare commented on spec/features/work_packages/tabs/relations_children_spec.rb:
```diff
@@ -113,4 +113,48 @@
+
+    before do
+      # Introspect FactoryBot to find all traits used to create work package custom fields
+      traits = FactoryBot.factories[:wp_custom_field].defined_traits
```
:yellow_circle: I like the idea of having each custom field type in a single test very much. It essentially mimics what one might do when testing this manually as well.

However, I am not a big fan of iterating over FactoryBot traits for that to be honest. The reason is that it's IMO easy for these to misalign with custom field types:
* I might create a custom field type, but forget to create a trait for it (e.g. the `multi_hierarchy` option that you just added in this PR)
    * no tests will fail, but this one will then be incomplete
* I might create a trait for the factory that makes sense, but would now (unknowingly) change this spec
    * if that makes this test fail it's annoying, but at least I will know
    * if that doesn't make it fail, then I wouldn't even notice that I changed this spec

Can't we instead use `OpenProject::CustomFieldFormat.available_formats` to iterate the custom field formats that are defined and then look for factory traits with a corresponding name?

This would make us iterate over an authoritative source of custom field types (one that _has_ to be changed when adding a new one) and it would as a side-effect also encourage to set-up proper factory traits when introducing a new type.

cbliard replied:
Actually, I used `OpenProject::CustomFieldFormat.available_formats` at first, and it did not work out of the box because some custom fields need additional options to be valid. For instance, `list` needs options, and `hierarchy` needs root item. Creating multi-value custom fields for types that support it was also a bit weird to code: calling `#multi_value_possible?` on a custom field instance to know if a multi-value version should be instantiated.

That's why I ended up using `FactoryBot`. I wanted something better, like selecting only traits which have redefined `field_format`, but did not manage to do it. That why it manually excludes the `multi_value` trait.

I still think that's an acceptable trade-off, as our factories should reflect what we have in code, and they should also work out of the box.

I share your concerns that they may be misaligned with the defined custom field types. What I can do is encapsulating these traits in a method, so that the logic can be reused somewhere else, and have that method tested, so that we continuously check that all custom field types with their multi-value variant are covered.

What do you think?

NobodysNightmare replied:
Your idea sounds good and would alleviate my concerns.

Let me quickly share what I had in mind, because this was probably not clear enough: I don't have a huge problem with assuming that certain factory traits exist, my main concern was that the factory bot traits were the _source_ of possible CF types to iterate from.

So if it helps, something along those lines would work for me, **too**:

```ruby
traits = OpenProject::CustomFieldFormat.available_formats.flat_map { |f| [f, "multi_#{f}".to_sym] }
traits.each do |trait|
  # what you already did before
end
```

cbliard replied:
Do you know what the format `"empty"` is used for?

cbliard replied:
Ah, found it in `config/initializers/custom_field_format.rb`

```ruby
  # This is an internal formatter used as a fallback in case a value is not found.
  # Setting the label to nil in order to avoid it becoming available for selection as a custom value format.
  fields.register OpenProject::CustomFieldFormat.new("empty",
                                                     label: nil,
                                                     order: 11,
                                                     formatter: "CustomValue::EmptyStrategy")
```

cbliard replied:
@NobodysNightmare I changed it the way we discussed. It now relies on `OpenProject::CustomFieldFormat` information to build up the traits list.

I also stumbled upon an order-dependent test. I fixed it.

Could you review it again please?

NobodysNightmare reviewed (changes requested):
Thanks for the quick fix. I had one comment on the way that the custom field types are iterated, let me know what you think :)

cbliard committed:
Fix order-dependent spec

Reproduction command: rspec --seed 10236 './spec/migrations/rename_visible_to_admin_only_in_custom_fields_spec.rb[1:1:1]' './spec/models/queries/projects/project_query_custom_field_order_spec.rb[1:5:1:1]'

cbliard committed:
Get custom fields traits from configuration instead of FactoryBot

Refactored some methods to use `OpenProject::CustomFieldFormat` to get
the `multi_value_possible` information.

NobodysNightmare reviewed (approved):
LGTM
